demos/coreHTTP/http_demo_s3_download_multithreaded.c

```diff
@@ -0,0 +1,1066 @@
+                                           const size_t xEnd )
+{
+    HTTPStatus_t xHTTPStatus = HTTPSuccess;
+    BaseType_t xStatus = pdPASS;
+
+    configASSERT( pxRequestInfo != NULL );
+
+    /* Set the buffer used for storing request headers. */
+    xRequestItem.xRequestHeaders.pBuffer = xRequestItem.ucHeaderBuffer;
+    xRequestItem.xRequestHeaders.bufferLen = democonfigUSER_BUFFER_LENGTH;
+
+    xHTTPStatus = HTTPClient_InitializeRequestHeaders( &( xRequestItem.xRequestHeaders ),
+                                                       pxRequestInfo );
+
+    if( xHTTPStatus != HTTPSuccess )
+    {
+        LogError( ( "Failed to initialize HTTP request headers: Error=%s.",
+                    HTTPClient_strerror( xHTTPStatus ) ) );
+        xStatus = pdFAIL;
+    }
+
+    if( xStatus == pdPASS )
+    {
+        xHTTPStatus = HTTPClient_AddRangeHeader( &( xRequestItem.xRequestHeaders ),
+                                                 xStart,
+                                                 xEnd );
+
+        if( xHTTPStatus != HTTPSuccess )
+        {
+            LogError( ( "Failed to add Range header to request headers: Error=%s.",
+                        HTTPClient_strerror( xHTTPStatus ) ) );
+            xStatus = pdFAIL;
+        }
+    }
+
+    if( xStatus == pdPASS )
+    {
+        /* Enqueue the request. */
+        LogInfo( ( "Enqueuing request for bytes %d to %d of S3 Object. ",
+                   ( int32_t ) xStart,
+                   ( int32_t ) xEnd ) );
+        LogDebug( ( "Request Headers:\n%.*s",
+                    ( int32_t ) xRequestItem.xRequestHeaders.headersLen,
+                    ( char * ) xRequestItem.xRequestHeaders.pBuffer ) );
+
+        xStatus = xQueueSendToBack( xRequestQueue,
+                                    &xRequestItem,
+                                    httpexampleDEMO_TICKS_TO_WAIT );
+
+        /* Ensure request was added to the queue. */
+        if( xStatus == pdFAIL )
+        {
+            LogError( ( "Could not enqueue request." ) );
+        }
+    }
+
+    return xStatus;
+}
+
+/*-----------------------------------------------------------*/
+
+static bool prvNotificationWaitLoop( uint32_t * pulNotification,
+                                     uint32_t ulExpectedBits,
+                                     bool xClearBits )
+{
+    uint32_t ulWaitCounter = 0U;
+    bool ret = true;
+
+    configASSERT( pulNotification != NULL );
+
+    while( ( *pulNotification & ulExpectedBits ) != ulExpectedBits )
+    {
+        xTaskNotifyWait( 0,
+                         ( xClearBits ) ? ulExpectedBits : 0,
+                         pulNotification,
+                         httpexampleDEMO_TICKS_TO_WAIT );
+
+        if( ++ulWaitCounter > httpexampleMAX_WAIT_ITERATIONS )
+        {
+            LogDebug( ( "Loop exceeded maximum wait time. Notification not received.\n" ) );
+            ret = false;
+            break;
+        }
```

Comment:
What is the purpose of this loop? Can't we add the total time to the timeout directly, if we need to wait?

Reply:
The loop was initially intended for task completion notifications, but since it's purpose was changed, the loop is extraneous -- I've removed it, thanks.